jwave/geometry.py

```diff
@@ -461,10 +463,102 @@ def __call__(self, p: Field, u: Field, rho: Field):
             return p.on_grid[self.positions[0]]
         elif len(self.positions) == 2:
             return p.on_grid[self.positions[0],
-                             self.positions[1]]    # type: ignore
+            self.positions[1]]  # type: ignore
         elif len(self.positions) == 3:
             return p.on_grid[self.positions[0], self.positions[1],
-                             self.positions[2]]    # type: ignore
+            self.positions[2]]  # type: ignore
+        else:
+            raise ValueError(
+                "Sensors positions must be 1, 2 or 3 dimensional. Not {}".
+                format(len(self.positions)))
+
+
+def _bli_function(x0, x, n, include_imag=False):
+    # x0 = position of sensors along axis
+    # x = grid
+    # n = size of grid
+    # include_imag = include the imaginary component?
+    dx = jnp.where((x - x0[:, None]) == 0, x - x0[:, None] + 1e-4, x - x0[:, None])  # Stop us getting nans.
```

Comment:
Can you help me understand what this `1e-4` term is here?

If that's to avoid division by zero when the point is on grid, we could perhaps do something similar to what is done to implement the `sinc` function in jax: https://github.com/google/jax/blob/6bc74d2a9874e1fe93a45191bb829c07dfee04fa/jax/_src/numpy/ufuncs.py#L717

The idea is to use a series representation when the input is zero (since the only thing we care of is to get the correct gradients).

If you don't have time to do this now no worries at all, just let me know and I will open an issue and tackle it afterwards.

Reply:
I was trying to do something like this: https://github.com/google/jax/issues/1052. I've changed line 481 to the following for clarity:

```python 
    dx = jnp.where((x - x0[:, None]) == 0, 1, x - x0[:, None])  # https://github.com/google/jax/issues/1052
```

Reply:
That's a very interesting sharp bit of jax that I didn't have the pleasure of seeing yet! 😄